### PR TITLE
 Redesign  the rest API

### PR DESCRIPTION
**Pull request by sbs2001**
TODOs 

1. Update documentation. ~Add tests specific to checking new API additions~
~2. Deal with inputting  packageurls as query param.~

### CODE REVIEW

File: vulnerabilities/models.py [hunk 1]
```diff
@@ -39,59 +39,87 @@ class Vulnerability(models.Model):
...
-        help_text='Summary of the vulnerability', blank=True)
-    cvss = models.FloatField(max_length=100, help_text='CVSS Score', null=True)
+
+    cve_id = models.CharField(max_length=50, help_text="CVE ID", unique=True, null=True)
+    summary = models.TextField(help_text="Summary of the vulnerability", blank=True)
+    cvss = models.FloatField(max_length=100, help_text="CVSS Score", null=True)
+
+    @property
+    def vulnerable_to(self):
+        qs = PackageRelatedVulnerability.objects.filter(
```
Comment (tdruez) [inline vulnerabilities/models.py:49]: Use Many2Many QuerySet syntax instead.

File: vulnerabilities/models.py [hunk 2]
```diff
@@ -39,59 +39,87 @@ class Vulnerability(models.Model):
...
     """
+
     vulnerabilities = models.ManyToManyField(
-        to='Vulnerability', through='PackageRelatedVulnerability')
+        to="Vulnerability", through="PackageRelatedVulnerability"
+    )
+
+    @property
+    def vulnerable_to(self):
+        qs = PackageRelatedVulnerability.objects.filter(
```
Comment (tdruez) [inline vulnerabilities/models.py:99]: Same as above, use Many2Many syntax.

File: vulnerabilities/models.py [hunk 3]
```diff
@@ -39,59 +39,87 @@ class Vulnerability(models.Model):
...
-            'Extra qualifying data for a package such as the name of an OS, '
-            'architecture, distro, etc.'
+            "Extra qualifying data for a package such as the name of an OS, "
+            "architecture, distro, etc."
         ),
-        null=True
+        null=True,
     )
 
     def set_package_url(self, package_url):
```
Comment (tdruez) [inline vulnerabilities/models.py:125]: Why are we overriding the default method from PackageURLMixin?

File: vulnerablecode/urls.py [hunk 4]
```diff
@@ -27,13 +27,17 @@
...
+# `DefaultRouter` requires `basename` when registering viewsets which don't
+# define a queryset.
+api_router.register(r"vulnerabilities", VulnerabilityViewSet, basename="vulnerability")
 
 
 urlpatterns = [
-    path('admin/', admin.site.urls),
-    re_path(r'^api/', include(api_router.urls))
+    path("admin/", admin.site.urls),
+    re_path(r"^api/", include(api_router.urls)),
```
Comment (tdruez) [inline vulnerablecode/urls.py:42]: What is the need for `re_path` instead of `path` here?

File: vulnerablecode/urls.py [hunk 5]
```diff
@@ -27,13 +27,17 @@
 from rest_framework.routers import DefaultRouter
 
 from vulnerabilities.api import PackageViewSet
+from vulnerabilities.api import VulnerabilityViewSet
 
 
-api_router = DefaultRouter()
-api_router.register(r'packages', PackageViewSet)
+api_router = DefaultRouter(trailing_slash=False)
```
Comment (tdruez) [inline vulnerablecode/urls.py:33]: What's the reasoning behind overriding the default trailing_slash convention?

**Review by tdruez — CHANGES_REQUESTED** [review]
The API code looks good.
See my comments for models and urls.

File: vulnerabilities/models.py (hunk 3 above)
Comment (sbs2001) [reply vulnerabilities/models.py:125]: This one uses JSONField the old`set_package_url` converts qualifiers into normalized string. More on this https://github.com/package-url/packageurl-python/pull/35#discussion_r434320271

File: vulnerablecode/urls.py (hunk 4 above)
Comment (sbs2001) [reply vulnerablecode/urls.py:42]: I haven't touched that, that's black doing it's thing.  Having said that  we can use `path` their alright.

File: vulnerablecode/urls.py (hunk 5 above)
Comment (sbs2001) [reply vulnerablecode/urls.py:33]: I had to discuss this with you guys. 

For the `ListView` using trailing_slash the endpoints look  odd. The endpoints with queries look like :
`/api/vulnerabilities/?vulnerability_id=foo`

After the override you get the :
`/api/vulnerabilities?vulnerability_id=foo`

Not sure this matter tho, so I might be missing something here.

Comment (tdruez) [reply vulnerablecode/urls.py:33]: I don't see the benefit of removing the trailing slash.

From https://www.django-rest-framework.org/api-guide/routers/

> Trailing slashes are conventional in Django, but are not used by default in some other frameworks such as Rails.

I think we should follow Django's conventions.

Comment (sbs2001) [reply vulnerablecode/urls.py:33]: :+1:

File: vulnerabilities/models.py [hunk 6]
```diff
@@ -39,75 +39,87 @@ class Vulnerability(models.Model):
...
+    cve_id = models.CharField(max_length=50, help_text="CVE ID", unique=True, null=True)
+    summary = models.TextField(help_text="Summary of the vulnerability", blank=True)
+    cvss = models.FloatField(max_length=100, help_text="CVSS Score", null=True)
 
     @property
     def vulnerable_to(self):
-        return self.packagerelatedvulnerability_set.filter(is_vulnerable=True)
+        return self.package_set.filter(
+            packagerelatedvulnerability__is_vulnerable=True,
+            packagerelatedvulnerability__vulnerability_id=self.id,
```
Comment (tdruez) [inline vulnerabilities/models.py:51]: I don't think this lookup is necessary since the related manager is starting on `self`.

**Review by tdruez — CHANGES_REQUESTED** [review]
Looks good. Although I don't think those lookups are necessary:
`packagerelatedvulnerability__vulnerability_id=self.id,`
and 
`packagerelatedvulnerability__package_id=self.id,`

Please confirm and update accordingly.

**Review by tdruez — APPROVED** [review]
LGTM!

**Review by pombredanne — APPROVED** [review]
LGTM :+1:

**Pull request merged**